availability-store/src/worker.rs

```diff
@@ -0,0 +1,989 @@
+
+	fn import_block(
+		&mut self,
+		block: BlockImportParams<Block>,
+		new_cache: HashMap<CacheKeyId, Vec<u8>>,
+	) -> Result<ImportResult, Self::Error> {
+		trace!(
+			target: LOG_TARGET,
+			"Importing block #{}, ({})",
+			block.header.number(),
+			block.post_header().hash()
+		);
+
+		if let Some(ref extrinsics) = block.body {
+			let relay_parent = *block.header.parent_hash();
+			let parent_id = BlockId::hash(*block.header.parent_hash());
+			// Extract our local position i from the validator set of the parent.
+			let validators = self.client.runtime_api().validators(&parent_id)
+				.map_err(|e| ConsensusError::ChainLookup(e.to_string()))?;
+
+			let our_id = self.our_id(&validators);
+
+			// Use a runtime API to extract all included erasure-roots from the imported block.
+			let candidates = self.client.runtime_api().get_heads(&parent_id, extrinsics.clone())
+				.map_err(|e| ConsensusError::ChainLookup(e.to_string()))?;
+
+			match candidates {
+				Some(candidates) => {
+					match our_id {
+						Some(our_id) => {
+							trace!(
+								target: LOG_TARGET,
+								"Our validator id is {}, the candidates included are {:?}",
+								our_id, candidates
+							);
+
+							for candidate in &candidates {
+								// If we don't yet have our chunk of this candidate,
+								// tell the worker to listen for one.
+								if self.availability_store.get_erasure_chunk(
+									&relay_parent,
+									candidate.block_data_hash,
+									our_id as usize,
+								).is_none() {
+									let msg = WorkerMsg::ListenForChunks(ListenForChunks {
+										relay_parent,
+										candidate_hash: candidate.hash(),
+										index: our_id as u32,
+										result: None,
+									});
+
+									let _ = self.to_worker.unbounded_send(msg);
+								}
+							}
+
+							let erasure_roots: Vec<_> = candidates
+								.iter()
+								.map(|c| c.erasure_root)
+								.collect();
+
+							// Inform the worker about new (relay_parent, erasure_roots) pairs
+							let (s, _) = oneshot::channel();
+							let msg = WorkerMsg::ErasureRoots(ErasureRoots {
+								relay_parent,
+								erasure_roots,
+								result: s,
+							});
+
+							let _ = self.to_worker.unbounded_send(msg);
+
+							let (s, _) = oneshot::channel();
+
+							// Inform the worker about the included parachain blocks.
+							let msg = WorkerMsg::ParachainBlocks(ParachainBlocks {
+								relay_parent,
+								blocks: candidates.into_iter().map(|c| (c, None)).collect(),
+								result: s,
+							});
+
+							let _ = self.to_worker.unbounded_send(msg);
+						}
+						None => {
+							trace!(
+								target: LOG_TARGET,
+								"Not in the validator set at parent block {}", parent_id
+							);
+						}
+					}
+				}
+				None => {
+					trace!(
+						target: LOG_TARGET,
+						"No parachain heads were included in block {}", block.header.hash()
+					);
+				},
+			}
+		}
+
+		self.inner.import_block(block, new_cache).map_err(Into::into)
```

Comment:
This is racing against the background worker handling the messages sent above, so import notifications might come out before the information is registered in the availability store. Not sure what to do about that

Reply:
I guess it's fine as long as we implement the network protocol in the end by getting notifications directly from the availability store.